### PR TITLE
Use Clang.jl to generate interface

### PR DESCRIPTION
**Pull request by kdheepak**
This PR uses [Clang.jl](https://github.com/JuliaInterop/Clang.jl/) to generate the interface to Xpress.

```
$ tree src/
src
├── api.jl
├── common.jl
├── ctypes.jl
├── lib.jl
├── MOIWrapper.jl
├── utils.jl
├── Xpress.jl
├── XpressSolverInterface.jl
├── xprs_attrs.jl
├── xprs_callbacks.jl
├── xprs_common.jl
├── xprs_constrs.jl
├── xprs_env.jl
├── xprs_full_defines.jl
├── xprs_highlevel.jl
├── xprs_iis.jl
├── xprs_model.jl
├── xprs_params.jl
├── xprs_quad.jl
├── xprs_solve.jl
├── xprs_userlic.jl
└── xprs_vars.jl
```

Clang.jl generates `ctypes.jl`, `common.jl` and `lib.jl`. The script used to generate this is in the `scripts` folder. These are all wrapped in a [`Lib`](https://github.com/kdheepak/Xpress.jl/blob/9464f198ef33c02ae6381fce8e72a72d5364dec1/src/Xpress.jl#L65-L78) module.

`api.jl` contains the user interface functions for the Xpress library functions. This can be used to provide a more Julian interface (see [`XPRSgetversion()`](https://github.com/kdheepak/Xpress.jl/blob/9464f198ef33c02ae6381fce8e72a72d5364dec1/src/api.jl#L34-L37) as an example).

The advantage of this approach is that for a new version of Xpress, we can re-run the `generate.jl` script and update everything in `Lib`, make minimal changes to `api.jl` but not need to change the function names in `api.jl`.

- [x] Generate `lib.jl`, `common.jl`, `ctypes.jl` using Clang.jl
- [ ] Write Julian `api.jl` wrapper for all functions in `Lib` module
- [ ] Write MOI wrapper on top of functions exposed in `api.jl`
- [ ] Documentation
- [ ] Tests

### CODE REVIEW

File: Project.toml [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+authors = ["Dheepak Krishnamurthy <me@kdheepak.com>"]
+version = "0.9.1"
+
+[deps]
+CEnum = "fa961155-64e5-5f13-b03f-caf6b980ea82"
+Libdl = "8f399da3-3557-5675-b5ff-fb832c97cbdb"
+
+[compat]
+CEnum = "^0.2"
+DocStringExtensions = "^0.8"
```
Comment (joaquimg) [inline Project.toml]: Do we really need this?

File: deps/deps.jl [hunk 2]
```diff
@@ -0,0 +1,22 @@
+import Libdl
+
+libname = string(Sys.iswindows() ? "" : "lib", "xprs", ".", Libdl.dlext)
+
+if !haskey(ENV, "XPRESSDIR")
```
Comment (joaquimg) [inline deps/deps.jl]: I don't think we should enforce this.
Personally, I don't use it.

File: deps/build.jl [hunk 3]
```diff
@@ -7,15 +7,6 @@ if isfile(depsfile)
     rm(depsfile)
 end
 
-function write_depsfile(path)
```
Comment (joaquimg) [inline deps/build.jl]: Why not use the deps file? its pretty standard

File: scripts/generate.jl [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+                                                                               # "xprd.h",
+                                                                               # "xprm_mc.h",
+                                                                               # "xprm_ni.h",
+                                                                               # "xprm_rt.h",
+                                                                               # "xprnls.h",
+                                                                               "xprs.h",
+                                                                               # "xprs_mse_defaulthandler.h",
+                                                                               # "XPRSdefaultMipSolEnumHandler.java",
+                                                                               # "xslp.h",
+                                                                            ]
```
Comment (joaquimg) [inline scripts/generate.jl]: do you have access do a xprl.h file?

File: Project.toml (hunk 1 above)
Comment (kdheepak) [reply Project.toml]: I've previously used it for API documentation in other packages and found it useful. So you can add a minimal docstring in the [source code](https://github.com/GMLC-TDC/HELICS.jl/blob/f421dc45df489ed1fb87a41d67d957f046a7a1f7/src/api.jl#L839-L844) and it will generate the [function signature in the documentation](https://gmlc-tdc.github.io/HELICS.jl/latest/api/#HELICS.helicsGetVersion-Tuple{}). I can remove it if you'd like to not have this dependency.

File: src/Xpress.jl [hunk 5]
```diff
@@ -1,99 +1,112 @@
...
+
+    # ## xprs_solve
+    # optimize, computeIIS, get_solution,
+    # get_status, OptimInfo, get_optiminfo, get_objval
+
+
+    ### include source files
+     module Lib
+
+        using CEnum
```
Comment (joaquimg) [inline src/Xpress.jl]: where is it used?

File: deps/build.jl (hunk 3 above)
Comment (kdheepak) [reply deps/build.jl]: Since in this case Xpress will be installed on the users machine and since we are not using `BinaryProvider`, I thought I'd commit the `deps` file to the repository. We will have to update the file manually anytime anything changes. So it's either use the `write_depsfile` function to create a `deps.jl` file on the user's machine or commit it to the repo. I don't feel strongly about this I can revert this change.

File: Project.toml (hunk 1 above)
Comment (joaquimg) [reply Project.toml]: I'd say that we should not use packages that are not used by JuMP, MOI, CPLEX and Gurobi.
Better for robusteness.
FICO's manual is online so, at leat most o the C API is covered

**Comment by joaquimg** [comment]
Nice to-do list.
I'd say we can sort of skip: "Write Julian api.jl wrapper for all functions in Lib module"
And only add there what is required for MOI.
This will solve most use cases and we can do the rest incrementally.

File: scripts/generate.jl (hunk 4 above)
Comment (kdheepak) [reply scripts/generate.jl]: No, I don't have an `xprl.h` file. The above files are the only ones in my install directory.

File: src/Xpress.jl (hunk 5 above)
Comment (kdheepak) [reply src/Xpress.jl]: Currently it doesn't appear to be used, but this is what Clang.jl uses if there are `enum`'s in the C header files. I'm assuming they will be required somewhere for flags or settings. If not, I'll remove it.

When this PR is close to ready, I'll go through, strip all unused code / dependencies, rebase into a small number of meaningful commits and then ping for a final review.

**Comment by kdheepak** [comment]
> I'd say we can sort of skip: "Write Julian api.jl wrapper for all functions in Lib module"

Sounds good! I'll prioritize making the necessary functions work.

File: deps/build.jl (hunk 3 above)
Comment (joaquimg) [reply deps/build.jl]: lets rever for now. to keep it similar to gurobi and cplex.

File: scripts/generate.jl (hunk 4 above)
Comment (joaquimg) [reply scripts/generate.jl]: ok, I will look into it

File: deps/deps.jl (hunk 2 above)
Comment (kdheepak) [reply deps/deps.jl]: Resolved in https://github.com/JuliaOpt/Xpress.jl/pull/43/commits/733a0770e340d71e0cee8848fe63705eeb5fe28a.

File: deps/build.jl (hunk 3 above)
Comment (kdheepak) [reply deps/build.jl]: Resolved in https://github.com/JuliaOpt/Xpress.jl/pull/43/commits/733a0770e340d71e0cee8848fe63705eeb5fe28a.

File: Project.toml (hunk 1 above)
Comment (kdheepak) [reply Project.toml]: Resolved in https://github.com/JuliaOpt/Xpress.jl/pull/43/commits/42319d303bae7cfbd62e76e0c2a19a0600f895cb.

File: src/Xpress.jl [hunk 6]
```diff
@@ -1,99 +1,119 @@
...
+    # include("xprs_highlevel.jl")
+
+    # include("xprs_solve.jl")
+    # include("xprs_callbacks.jl")
+    # include("xprs_iis.jl")
+
+    # include("XpressSolverInterface.jl")
+    # include("MOIWrapper.jl")
+
+    function __init__()
```
Comment (joaquimg) [inline src/Xpress.jl]: This is not related to the re write.
But I was thinking about removing this and forcing the user to call the functions explicitly.
One intermediary option would be to keep a global Vector variable “is_initialized” That is set by init. In this case the mains proble constructor could call it if false.

File: src/api.jl [hunk 7]
```diff
@@ -0,0 +1,10962 @@
...
+prob1 to
+prob2:
+
+*Related*
+
+XPRScopycallbacks,
+XPRScopycontrols,
+XPRScreateprob.
+"""
+function copyprob(dest, src, probname)
```
Comment (joaquimg) [inline src/api.jl]: Nice docstrings

File: src/Xpress.jl (hunk 6 above)
Comment (kdheepak) [reply src/Xpress.jl]: You mean `XPRSinit`, right? I prefer that too. In fact in my local copy I don't have that and the license check in the `__init__`, but I just don't commit that patch.

Tradeoffs as I see it: 

- We need to have XPRSinit called and run successfully before anyone can use this package. Currently someone can run `using Xpress`, have it fail and type `using Xpress` again, and since Julia already ran `__init__` it won't run it again. This might give the impression to a user that the package initialized successfully when they import the second time. I've been in this situation. If anything, we can call it in `Xpress` module scope. If it errors, typing `using Xpress` again will force precompilation.
- If we don't call XPRSinit during import or initialization any time a user calls any Xpress library function we should throw the appropriate error. So we have to run a check before every function call (I don't like this), or run a check when an error is thrown (I prefer this).

File: src/api.jl [hunk 8]
```diff
@@ -0,0 +1,10962 @@
...
+PRESOLVE to
+0, turning off the presolve facility prior to optimization:
+
+*Related*
+
+XPRSgetintcontrol,
+XPRSsetdblcontrol,
+XPRSsetstrcontrol.
+"""
+function setintcontrol(prob::XpressProblem, _index, _ivalue)
```
Comment (joaquimg) [inline src/api.jl]: Clang.jl has an option to enforce type?
thats something I have been thinking about for a while... because these C functions can easily (and silently) break if types are wrong

File: src/api.jl [hunk 9]
```diff
@@ -0,0 +1,10962 @@
+"""
```
Comment (joaquimg) [inline src/api.jl]: this file is not auto generated?

File: src/Xpress.jl (hunk 6 above)
Comment (joaquimg) [reply src/Xpress.jl]: Lets go for: not calling on initialization.
This "run a check when an error is thrown (I prefer this)." is reasonable.
But I think almost all functions depend on a `XPRSprob`, we might consider testing the initialization inside that constructor.
So maybe check in constructor and when error is thrown?

File: src/api.jl (hunk 9 above)
Comment (kdheepak) [reply src/api.jl]: I generated it using a script initially, and have been adding stuff to it manually now. If I need to edit all the functions I either edit it with another script or just using editor magic. This can probably be done better, but at the moment I don't know what patterns exist and what can be abstracted so I'm just going down the dumb and verbose path.

File: src/api.jl (hunk 8 above)
Comment (kdheepak) [reply src/api.jl]: Clang.jl does have an [option](https://github.com/JuliaInterop/Clang.jl/blob/7823753f1c5c77a4678f0f33f92b03520962d044/src/wrap_c.jl#L48) but it's [not exposed in their wrapper convenience functions](https://github.com/JuliaInterop/Clang.jl/blob/03ddf229023fd17ea96c61597bfbea73581a73df/src/compat.jl#L33-L47). 

I ran it on this project to show as an example what it does. This is just a sample of the changes to `src/lib.jl`.

```
diff --git a/src/lib.jl b/src/lib.jl
index b7fb2cf..0704c8e 100644
--- a/src/lib.jl
+++ b/src/lib.jl
@@ -2,15 +2,15 @@
 # Automatically generated using Clang.jl


-function XPRScopycallbacks(dest, src)
+function XPRScopycallbacks(dest::XPRSprob, src::XPRSprob)
     ccall((:XPRScopycallbacks, libxprs), Cint, (XPRSprob, XPRSprob), dest, src)
 end

-function XPRScopycontrols(dest, src)
+function XPRScopycontrols(dest::XPRSprob, src::XPRSprob)
     ccall((:XPRScopycontrols, libxprs), Cint, (XPRSprob, XPRSprob), dest, src)
 end

-function XPRScopyprob(dest, src, probname)
+function XPRScopyprob(dest::XPRSprob, src::XPRSprob, probname)
     ccall((:XPRScopyprob, libxprs), Cint, (XPRSprob, XPRSprob, Cstring), dest, src, probname)
 end

@@ -18,7 +18,7 @@ function XPRScreateprob(_probholder)
     ccall((:XPRScreateprob, libxprs), Cint, (Ptr{XPRSprob},), _probholder)
 end

-function XPRSdestroyprob(_prob)
+function XPRSdestroyprob(_prob::XPRSprob)
     ccall((:XPRSdestroyprob, libxprs), Cint, (XPRSprob,), _prob)
 end

@@ -42,7 +42,7 @@ function XPRSgetdaysleft(daysleft)
     ccall((:XPRSgetdaysleft, libxprs), Cint, (Ptr{Cint},), daysleft)
 end

-function XPRSsetcheckedmode(checked_mode)
+function XPRSsetcheckedmode(checked_mode::Cint)
     ccall((:XPRSsetcheckedmode, libxprs), Cint, (Cint,), checked_mode)
 end

@@ -62,1123 +62,1123 @@ function XPRSendlicensing()
     ccall((:XPRSendlicensing, libxprs), Cint, ())
 end

-function XPRSgetlicerrmsg(msg, len)
+function XPRSgetlicerrmsg(msg, len::Cint)
     ccall((:XPRSgetlicerrmsg, libxprs), Cint, (Cstring, Cint), msg, len)
 end

-function XPRSsetlogfile(prob, logname)
+function XPRSsetlogfile(prob::XPRSprob, logname)
     ccall((:XPRSsetlogfile, libxprs), Cint, (XPRSprob, Cstring), prob, logname)
 end

-function XPRSsetintcontrol(prob, _index, _ivalue)
+function XPRSsetintcontrol(prob::XPRSprob, _index::Cint, _ivalue::Cint)
     ccall((:XPRSsetintcontrol, libxprs), Cint, (XPRSprob, Cint, Cint), prob, _index, _ivalue)
 end

-function XPRSsetintcontrol64(prob, _index, _ivalue)
+function XPRSsetintcontrol64(prob::XPRSprob, _index::Cint, _ivalue::Clong)
     ccall((:XPRSsetintcontrol64, libxprs), Cint, (XPRSprob, Cint, Clong), prob, _index, _ivalue)
 end

-function XPRSsetdblcontrol(prob, _index, _dvalue)
+function XPRSsetdblcontrol(prob::XPRSprob, _index::Cint, _dvalue::Cdouble)
     ccall((:XPRSsetdblcontrol, libxprs), Cint, (XPRSprob, Cint, Cdouble), prob, _index, _dvalue)
 end

-function XPRSinterrupt(prob, reason)
+function XPRSinterrupt(prob::XPRSprob, reason::Cint)
     ccall((:XPRSinterrupt, libxprs), Cint, (XPRSprob, Cint), prob, reason)
 end

-function XPRSgetprobname(prob, _svalue)
+function XPRSgetprobname(prob::XPRSprob, _svalue)
     ccall((:XPRSgetprobname, libxprs), Cint, (XPRSprob, Cstring), prob, _svalue)
 end

-function XPRSgetqobj(prob, _icol, _jcol, _dval)
+function XPRSgetqobj(prob::XPRSprob, _icol::Cint, _jcol::Cint, _dval)
     ccall((:XPRSgetqobj, libxprs), Cint, (XPRSprob, Cint, Cint, Ptr{Cdouble}), prob, _icol, _jcol, _dval)
 end

-function XPRSsetprobname(prob, _svalue)
+function XPRSsetprobname(prob::XPRSprob, _svalue)
     ccall((:XPRSsetprobname, libxprs), Cint, (XPRSprob, Cstring), prob, _svalue)
 end
:
```

There's more changes but I think this diff is sufficient for this discussion. Let me know if you want to see the full file, I can commit and revert.

This is strictly typed for some arguments (`Cint` and `Ptr{Nothing}`) but not others (`CString`, `Ptr{Cint}` etc). I think we do need strictly typed functions that users call but it shouldn't be the functions in `src/lib.jl`.

For example, the following code:

```
julia> using Xpress
[ Info: Precompiling Xpress [9e70acf3-d6c9-5be6-b5bd-4e2c73e3e054]
[ Info: Xpress: Found license file /Users/USER/local/xpressmp/bin/xpauth.xpr
[ Info: Xpress: Development license detected.

julia> xp = Xpress.XpressProblem()
Xpress.XpressProblem(Ptr{Nothing} @0x000000013ae14000)

julia> Xpress.getprobname(xp)
ERROR: MethodError: no method matching XPRSgetprobname(::Xpress.XpressProblem, ::Cstring)
Closest candidates are:
  XPRSgetprobname(::Ptr{Nothing}, ::Any) at /Users/USER/GitRepos/Xpress.jl/src/lib.jl:90
Stacktrace:
 [1] invoke(::Function, ::Int64, ::Type{String}, ::Xpress.XpressProblem) at /Users/USER/GitRepos/Xpress.jl/src/utils.jl:21
 [2] getprobname(::Xpress.XpressProblem) at /Users/USER/GitRepos/Xpress.jl/src/api.jl:593
 [3] top-level scope at REPL[5]:1

julia>
```

doesn't work with this strictly typed version of `src/lib.jl`. 

But with the current latest commit on this branch:

```
julia> using Xpress
[ Info: Precompiling Xpress [9e70acf3-d6c9-5be6-b5bd-4e2c73e3e054]
[ Info: Xpress: Found license file /Users/USER/local/xpressmp/bin/xpauth.xpr
[ Info: Xpress: Development license detected.

julia> xp = Xpress.XpressProblem()
Xpress.XpressProblem(Ptr{Nothing} @0x000000012e0b5000)

julia> Xpress.getprobname(xp)
""

julia>
```

it works fine. This is because I've created a [`struct XpressProblem`](https://github.com/JuliaOpt/Xpress.jl/pull/43/files#diff-4cd08f5188a8e8524aa70a7f07808f0aR42-R63):

```
abstract struct CWrapper end
struct XpressProblem <: CWrapper
    ptr::Lib.XPRSprob
end
```

that just holds the `Ptr{Nothing}` to the Xpress problem and I've defined the following `unsafe_convert` on it:

```
Base.unsafe_convert(T::Type{Ptr{Nothing}}, t::CWrapper) = t.ptr
```

`unsafe_convert` is called during `ccall`.

This function (`getprobname`) is strictly typed in `src/api.jl` and only allows users to pass in a type we have defined (`XpressProblem`). We can define additional functions on `XpressProblem`, for example some nicer `show` functionality.

The reason the strictly typed version does not work is that it requires the caller to pass in `Ptr{Nothing}`. Which means we can't use `convert` or `unsafe_convert`. Which also means we will still need to use a `XpressProblem` like `Ptr{Nothing}` holder and pass `p.ptr` explicitly. This means we have to strictly type `src/api.jl` AND `src/lib.jl`.

We will also have to convert do a lot of the manual conversion of types e.g `Int64` to `Cint` etc, which Julia does for us automatically during the `ccall` for us now. And we still have to manage conversions for `Cstring` and `Ptr{Int}` as we are doing now anyway. So I don't see any benefit to making the functions have strict types in `src/lib.jl`.

I completely agree with your point though, that strictly typed is better when interfacing with C code. We don't want segfaults. I think the best way to go about it is to make `src/lib.jl` not strict but make `src/api.jl` as strict as possible. It's more manual work though (which I'm prepared to do). I've done it in [other](https://github.com/dss-extensions/OpenDSSDirect.jl) [projects](https://github.com/GMLC-TDC/HELICS.jl) and I think it works well in providing a nice Julian interface (more intuitive calling for users, more explicit error messages when things go wrong, etc). My only concern with the current approach is that it is a lot of code repetition but I can't see a better way to provide a nice interface.

Thoughts? I'm open to feedback if you think I've missed something.

File: src/api.jl (hunk 9 above)
Comment (joaquimg) [reply src/api.jl]: I see, since we want to be extensive, we might consider automating that in the future. Ok lets keep it that way for now

File: src/Xpress.jl (hunk 6 above)
Comment (kdheepak) [reply src/Xpress.jl]: Resolved in https://github.com/JuliaOpt/Xpress.jl/pull/43/commits/8498ec359d03f369b31c2bf6fcd0143af82d59ab.

```
$ julia --project
               _
   _       _ _(_)_     |  Documentation: https://docs.julialang.org
  (_)     | (_) (_)    |
   _ _   _| |_  __ _   |  Type "?" for help, "]?" for Pkg help.
  | | | | | | |/ _` |  |
  | | |_| | | | (_| |  |  Version 1.3.0-rc4.1 (2019-10-15)
 _/ |\__'_|_|_|\__'_|  |  Official https://julialang.org/ release
|__/                   |

julia> using Xpress
[ Info: Precompiling Xpress [9e70acf3-d6c9-5be6-b5bd-4e2c73e3e054]

julia> Xpress.getversion()
v"29.1.10"

julia> Xpress.getbanner()
ERROR - call to XPRS function when global environment not initialised - please call XPRSinit() before attempt to call Xpress-Optimizer functions.
ERROR: XpressError: Unrecoverable error. Unable to invoke XPRSgetbanner
Stacktrace:
 [1] invoke(::Function, ::Int64, ::Type{String}) at /Users/USER/GitRepos/Xpress.jl/src/utils.jl:22
 [2] getbanner() at /Users/USER/GitRepos/Xpress.jl/src/api.jl:257
 [3] top-level scope at REPL[3]:1

julia> xp = Xpress.XpressProblem()
ERROR - call to XPRS function when global environment not initialised - please call XPRSinit() before attempt to call Xpress-Optimizer functions.
ERROR: Unable to create a Xpress Problem. Received error code 255
Stacktrace:
 [1] error(::String) at ./error.jl:33
 [2] Xpress.XpressProblem() at /Users/USER/GitRepos/Xpress.jl/src/helper.jl:56
 [3] top-level scope at REPL[4]:1

julia> Xpress.initialize()
[ Info: Xpress: Found license file /Users/USER/local/xpressmp/bin/xpauth.xpr
[ Info: Xpress: Development license detected.

julia> Xpress.getbanner()
"FICO Xpress-Optimizer 64-bit v29.01.10 (Hyper capacity)\n(c) Copyright Fair Isaac Corporation 1983-2016. All rights reserved"

julia> xp = Xpress.XpressProblem()
Xpress.XpressProblem(Ptr{Nothing} @0x00000001125f5000)

julia>


```

File: src/api.jl (hunk 8 above)
Comment (joaquimg) [reply src/api.jl]: Thanks for all the info.

I agree that Clang's types don't seem very useful. The need for a `CString` workaround and loosing the auto conversion for `Int64` to `Cint` are good examples of the problem.

Code repetition doesn't look that bad, I am afraid of the manual generation part. If it was automatically generated it could work well....

Brainstorming now:

About types...

What types would we need in `api.jl`?
`Cint` <- `Int32` and `Int64`
`Cdouble` <- `Float64`
`Cstring` < -`String` (maybe some `AbstractString` ?)
`Ptr{Cvoid}` <- `Ptr{Nothing}` / `XpressProblem`
`Ptr{Cint}` <- `Vector{Int32}`, `Vector{Int64}` and `Vector{Nothing}` (because we might need it to be C_NULL)
`Ptr{Cdouble}` <- `Vector{Float64}` and `Vector{Nothing}`

Looks doable to automated in `api.jl`?

Comment (kdheepak) [reply src/api.jl]: I would say yes it's possible, but not straightforward to cover all ways C code needs to be wrapped.

For example, take `getlicerrmsg`. In `src/lib.jl` its arguments are [`XPRSgetlicerrmsg(::Cstring, ::Cint)`](https://github.com/JuliaOpt/Xpress.jl/pull/43/files#diff-d7b1ad6cb56b02c302a7572a14873d38R65-R67) but in `src/api.jl` it is [`getlicerrmsg()`](https://github.com/JuliaOpt/Xpress.jl/pull/43/files#diff-9134b69774e09f102e883a13c38c0ab9R416-R421).

Another example is functions like `XPRSgetprobname` and `XPRSsetprobname`. Both have type signatures as `Tuple{Ptr{Nothing}, Cstring}`. But `getprobname` takes `xp::XpressProblem` and returns `String` and `setprobname` takes `xp::XpressProblem` and `name::String` and returns `nothing::Nothing`.

I guess we can automate this based on the names (`get`/`set`) etc but I need to look into all the functions to know if there are other nuances.

I think the main thing I'm aiming for now is that in `src/api.jl` we should have as minimal contents in the body of each function as possible. It should be easy to understand for people so that we can easily debug / extend and potentially automate in the future. Currently my plan is to do that using macros. See [getprobname](https://github.com/JuliaOpt/Xpress.jl/pull/43/files#diff-9134b69774e09f102e883a13c38c0ab9R593) as an example.

Comment (joaquimg) [reply src/api.jl]: So, first generate Clang's `lib.jl` the post processes that with protected types

Comment (joaquimg) [reply src/api.jl]: Ok...

> ( I have another script that gets the docstrings from the xpress documentation website and adds them to the appropriate function in a julia file )

Please add the automation files to the git repo.

> If api.jl is strictly typed for all input arguments, I don't see a benefit to doing this though

The plan is to enforce type with `Union`s, right? so that the user can call with both `Int32` and `Int64`

> I think I'm going to focus on getting api.jl working first so that work on the MOI wrapper can continue

Lets do it! Lets get MOI up and running. You are planning to work on the MOI_wrapper in this same PR, right?

Comment (jd-lara) [reply src/api.jl]: Yeah, we are working on both in close coordination

File: src/helper.jl [hunk 10]
```diff
@@ -0,0 +1,436 @@
...
+n_special_ordered_sets(prob::XpressProblem) = getintattrib(prob, Lib.XPRS_SETS)
+n_quadratic_constraints(prob::XpressProblem) = getintattrib(prob, Lib.XPRS_QCONSTRAINTS)
+n_non_zero_elements(prob::XpressProblem) = getintattrib(prob, Lib.XPRS_ELEMS)
+n_quadratic_elements(prob::XpressProblem) = getintattrib(prob, Lib.XPRS_QELEMS)
+n_quadratic_row_coefficients(prob::XpressProblem) = getintattrib(prob, Lib.XPRS_QCELEMS)
+n_entities(prob::XpressProblem) = getintattrib(prob, Lib.XPRS_MIPENTS)
+n_setmembers(prob::XpressProblem) = getintattrib(prob, Lib.XPRS_SETMEMBERS)
+
+n_original_variables(prob::XpressProblem) = getintattrib(prob, Lib.XPRS_ORIGINALCOLS)
+n_original_constraints(prob::XpressProblem) = getintattrib(prob, Lib.XPRS_ORIGINALROWS)
```
Comment (kdheepak) [inline src/helper.jl]: I picked names for these functions based on the Xpress documentation and what I thought would be most readable. I'm open to other suggestions or using what was previously used if people find this too verbose.

File: src/Xpress.jl (hunk 5 above)
Comment (kdheepak) [reply src/Xpress.jl]: Resolved in 44215b9.

File: src/helper.jl (hunk 10 above)
Comment (joaquimg) [reply src/helper.jl]: those are fine, verbose is good

File: src/helper.jl [hunk 11]
```diff
@@ -59,7 +59,7 @@ mutable struct XpressProblem <: CWrapper
         ptr = ref[]
         @assert ptr != C_NULL "Failed to create XpressProblem. Received null pointer from Xpress C interface."
         p = new(ptr, Any[], 0.0)
-        finalizer((p) -> destroyprob(p), p)
+        atexit(() -> destroyprob(p))
```
Comment (joaquimg) [inline src/helper.jl]: if I understand correctly, `atexit` is only called at the `exit` command, while `finalizer` is called at garbage collection.
Correct me if I am wrong.
If I am right, we will acumulate problems during execution and only get rid of them when the program finishes, this would lead to huge memory issues.

Comment (kdheepak) [reply src/helper.jl]: I meant to add a `TODO` comment here. 

Yes, you are correct that functions registered through `finalizer` are called when the objects are garbage collected. However, functions registered with [`atexit`](https://docs.julialang.org/en/v1/base/base/#Base.atexit) are run before object finalizers. This means that if we use `finalizer` in the struct constructor, we cannot register the [`free` function to run during `atexit`](https://github.com/JuliaOpt/Xpress.jl/blob/7bb86ec04036a07697dcd934ecb646685b0dec28/src/Xpress.jl#L90) when we call `Xpress.initialize()`. If we do use `finalizer` in the inner constructor to call `destroyprob` AND use `atexit` during `initialize` to call `free`, then:

```
julia --project
               _
   _       _ _(_)_     |  Documentation: https://docs.julialang.org
  (_)     | (_) (_)    |
   _ _   _| |_  __ _   |  Type "?" for help, "]?" for Pkg help.
  | | | | | | |/ _` |  |
  | | |_| | | | (_| |  |  Version 1.3.0-rc4.1 (2019-10-15)
 _/ |\__'_|_|_|\__'_|  |  Official https://julialang.org/ release
|__/                   |

julia> using Xpress
[ Info: Precompiling Xpress [9e70acf3-d6c9-5be6-b5bd-4e2c73e3e054]

julia> Xpress.initialize()
[ Info: Xpress: Found license file /Users/USER/local/xpressmp/bin/xpauth.xpr
[ Info: Xpress: Development license detected.

julia> xp = Xpress.XpressProblem()
Xpress Problem:
    type   : LP
    sense  : minimize
    number of variables                    = 0
    number of linear constraints           = 0
    number of quadratic constraints        = 0
    number of sos constraints              = 0
    number of non-zero coeffs              = 0
    number of non-zero qp objective terms  = 0
    number of non-zero qp constraint terms = 0
    number of integer entities             = 0


julia> exit()
ERROR - call to XPRS function when global environment not initialised - please call XPRSinit() before attempt to call Xpress-Optimizer functions.
error in running finalizer: Xpress.XpressError(errorcode=255, msg="Unable to call destroyprob")
```

we will get the above error during `exit()`.

I posted this question on slack and @simonbyrne suggested I look at [MathLink.jl](https://github.com/JuliaInterop/MathLink.jl/blob/7624db2d977990659f310103f2c7bf7d0266aa23/src/init.jl) and [MPI.jl](https://github.com/JuliaParallel/MPI.jl/blob/f471b62fea9ff46411f83556b95eac51023a9667/src/environment.jl) to see how they manage environments, and I think we can do something similar.

File: src/api.jl (hunk 9 above)
Comment (kdheepak) [reply src/api.jl]: After exploring Clang.jl source and documentation more, I believe we can autogenerate `api.jl` (and least most of it) directly from Clang.jl. I'll keep proceeding with the current implementation for now though. I will submit a separate PR at a later stage where all this is auto generated.

File: src/helper.jl (hunk 11 above)
Comment (joaquimg) [reply src/helper.jl]: I never had any problem by not freeing an xpress env, did you?

File: src/api.jl [hunk 12]
```diff
@@ -1941,7 +1941,7 @@ XPRSaddnames,
 XPRSloadglobal,
 XPRSloadqglobal.
 """
-function addsetnames(prob::XpressProblem, _sname, first, last)
+function addsetnames(prob::XpressProblem, _sname, first::Int, last::Int)
```
Comment (joaquimg) [inline src/api.jl]: maybe Integer instead of Int?

File: src/api.jl [hunk 13]
```diff
@@ -3477,8 +3484,10 @@ The following example retrieves the upper bounds for the columns of the current
 XPRSchgbounds,
 XPRSgetlb.
 """
-function getub(prob::XpressProblem, _dbdu, first, last)
+function getub(prob::XpressProblem, first::Int, last::Int)
+    _dbdu = Vector{Float64}(undef, last - first + 1)
```
Comment (joaquimg) [inline src/api.jl]: For functions that query solution I find it very useful to have in-place versions.
so that I keep a single allocated vector and just update it in-place, its very useful if you are solving many-many-many LP's in sequence.
For those function I would vote for two implementations:
```julia
function getub(prob::XpressProblem, _dbdu::Vector{Float64}, first::Int, last::Int)
    @assert length(_dbdu) = last - first + 1
    @checked Lib.XPRSgetub(prob, _dbdu, first, last)
    return nothing
end
function getub(prob::XpressProblem, first::Int, last::Int)
    _dbdu = Vector{Float64}(undef, last - first + 1)
    getub(prob, _dbdu, first, last)
    return _dbdu
end
```
It is not essencial for getub, but for getting primals and duals its a must.

File: src/api.jl [hunk 14]
```diff
@@ -4338,8 +4364,18 @@ XPRSchgrowtype,
...
-    @checked Lib.XPRSgetrowtype(prob, _srowtype, first, last)
+function getrowtype(prob::XpressProblem, first::Integer, last::Integer)
+    n_elems = last - first + 1
+    if n_elems <= 0
+        n_elems = n_constraints(prob)
+        first = 0
+        last = n_elems - 1
+    else
+        first = first - 1
+    end
```
Comment (joaquimg) [inline src/api.jl]: this coudl be a function:
`n_elems, first, last = myfunc(first, last)`

Comment (kdheepak) [reply src/api.jl]: Yes, agreed. I'll make the change. It would need to take a function argument too, since it can `get_marks(prob, first, last, n_constraints)` or `get_marks(prob, first, last, n_rows)`.

File: src/api.jl (hunk 12 above)
Comment (kdheepak) [reply src/api.jl]: Resolved in e5902ac.

File: src/helper.jl (hunk 11 above)
Comment (kdheepak) [reply src/helper.jl]: I have not, but I figured it exists for a reason and thought I should do it. I can experiment with an implementation that has both `destroyprob` and `free` and if I'm not able to come up with something that works and is clean, I'll just use what is in the current `master`, i.e. not use `free` and register `destroyprob` using `finalizer`.

Comment (joaquimg) [reply src/helper.jl]: nice, we leave the `free` thing as a TODO

**Comment by joaquimg** [comment]
Bump, any news or timeline?

What if we change the plan of this PR to:
- Generate lib.jl, common.jl, ctypes.jl using Clang.jl (done!)
- Write (minimal) Julian api.jl wrapper for all functions in Lib module
- Write (minimal) MOI wrapper (poor performance, only LP and simple MIP integer and binary) on top of functions exposed in api.jl (minimal version)
- Pass MOI (most basic LP and MIP) tests

This would allow Xpress to be used in the latest JuMP/MOI, which enable the broader public to use Xpress.

Then we could have a second (set of) PR(s) with:
- Write (more complete) Julian api.jl wrapper for all functions in Lib module
- Write (more complete) MOI wrapper (High Performance, QP, QCQP, SOCP, Callbacks, SOS, SemiContinuous etc) on top of functions exposed in api.jl (minimal version)
- Documentation
- Tests

This second could be done incrementally by many smaller PR's that people add as needed... and could be done in parallel.

**Comment by kdheepak** [comment]
I had a few things come up this week and also I'm likely not going to be able to work on this again till Thursday next week. 

But I'm on board with your plan! Currently in this PR some of the essential functions seems to be done in `api.jl` that wrap the corresponding C interface, now it is a question of what it takes to get the MOI interface working and to pass tests. @jd-lara and I will chat and report back.

**Comment by joaquimg** [comment]
I will give it try today

**Comment by jd-lara** [comment]
I will in NREL during the week of the 18 to wrap it up with Dheepak
